sql/colexec: TestCaseOpRandomized failed

sql/colexec.TestCaseOpRandomized [failed](https://mesolite.cluster.engflow.com/invocations/default/94109329-9db8-43f6-8803-76b977e8dbba?testReportRun=1&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NvbGV4ZWM6Y29sZXhlY190ZXN0) on master @ [0fd08b1c5cc129c2b1714a81ded82248628d8603](https://github.com/cockroachdb/cockroach/commits/0fd08b1c5cc129c2b1714a81ded82248628d8603):


```
=== RUN   TestCaseOpRandomized
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestCaseOpRandomized979110315
    test_log_scope.go:76: use -show-logs to present logs inline
    proj_utils_test.go:69: 
        	Error Trace:	pkg/sql/colexec/proj_utils_test.go:69
        	            				pkg/sql/colexec/case_test.go:162
        	Error:      	Received unexpected error:
        	            	processorConstructor is nil: unhandled cast to oid
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.opResult.createAndWrapRowSource
        	            	  | 	pkg/sql/colexec/colbuilder/execplan.go:536
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.opResult.wrapPostProcessSpec
        	            	  | 	pkg/sql/colexec/colbuilder/execplan.go:1887
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.NewColOperator
        	            	  | 	pkg/sql/colexec/colbuilder/execplan.go:1730
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils.CreateTestProjectingOperator
        	            	  | 	pkg/sql/colexec/colexectestutils/proj_utils.go:88
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.assertProjOpAgainstRowByRow
        	            	  | 	pkg/sql/colexec/proj_utils_test.go:66
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.TestCaseOpRandomized
        	            	  | 	pkg/sql/colexec/case_test.go:162
        	            	  | [...repeated from below...]
        	            	Wraps: (2) processorConstructor is nil
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase.init
        	            	  | 	bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecbase/cast.eg.go:77
        	            	  | runtime.doInit1
        	            	  | 	GOROOT/src/runtime/proc.go:7321
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:7288
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:254
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1700
        	            	Wraps: (4) unhandled cast to oid
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
        	Test:       	TestCaseOpRandomized
    panic.go:629: -- test log scope end --
--- FAIL: TestCaseOpRandomized (0.00s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCaseOpRandomized.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48093

sql/colexec.TestCaseOpRandomized [failed](https://tanzanite.cluster.engflow.com/invocations/default/ba4262cd-995c-40e0-842e-3a303f7ba58a?testReportRun=6&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NvbGV4ZWM6Y29sZXhlY190ZXN0) on master @ [8bd38972db32f5e683ba6dc0eb599334919919e5](https://github.com/cockroachdb/cockroach/commits/8bd38972db32f5e683ba6dc0eb599334919919e5):


```
=== RUN   TestCaseOpRandomized
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestCaseOpRandomized95268264
    test_log_scope.go:76: use -show-logs to present logs inline
    proj_utils_test.go:69: 
        	Error Trace:	pkg/sql/colexec/proj_utils_test.go:69
        	            				pkg/sql/colexec/case_test.go:162
        	Error:      	Received unexpected error:
        	            	processorConstructor is nil: unhandled cast to oid
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.opResult.createAndWrapRowSource
        	            	  | 	pkg/sql/colexec/colbuilder/execplan.go:536
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.opResult.wrapPostProcessSpec
        	            	  | 	pkg/sql/colexec/colbuilder/execplan.go:1887
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colbuilder.NewColOperator
        	            	  | 	pkg/sql/colexec/colbuilder/execplan.go:1730
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexectestutils.CreateTestProjectingOperator
        	            	  | 	pkg/sql/colexec/colexectestutils/proj_utils.go:88
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.assertProjOpAgainstRowByRow
        	            	  | 	pkg/sql/colexec/proj_utils_test.go:66
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec.TestCaseOpRandomized
        	            	  | 	pkg/sql/colexec/case_test.go:162
        	            	  | [...repeated from below...]
        	            	Wraps: (2) processorConstructor is nil
        	            	Wraps: (3) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecbase.init
        	            	  | 	bazel-out/k8-fastbuild/bin/pkg/sql/colexec/colexecbase/cast.eg.go:77
        	            	  | runtime.doInit1
        	            	  | 	GOROOT/src/runtime/proc.go:7321
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:7288
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:254
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1700
        	            	Wraps: (4) unhandled cast to oid
        	            	Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.leafError
        	Test:       	TestCaseOpRandomized
    panic.go:629: -- test log scope end --
--- FAIL: TestCaseOpRandomized (0.01s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=6</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCaseOpRandomized.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/colexec/proj_utils_test.go
+++ b/pkg/sql/colexec/proj_utils_test.go
@@ -30,17 +30,28 @@ import (
 // engine because all datum-backed types have the same backing datumVec which
 // would occur disproportionally often without adjusting the weights.
 func getRandomTypeFavorNative(rng *rand.Rand) *types.T {
-	typ := randgen.RandType(rng)
+	randTyp := func() *types.T {
+		for {
+			typ := randgen.RandType(rng)
+			switch typ.Family() {
+			case types.OidFamily:
+				// Skip the Oid family since casts to Oid type are handled by
+				// falling back to the row-by-row engine, and we don't set
+				// ProcessorConstructor in projection tests.
+			case types.VoidFamily:
+				// Skip the void family because it doesn't have some basic
+				// comparison operators defined.
+			default:
+				return typ
+			}
+		}
+	}
+	typ := randTyp()
 	for retry := 0; retry < 3; retry++ {
 		if typeconv.TypeFamilyToCanonicalTypeFamily(typ.Family()) != typeconv.DatumVecCanonicalTypeFamily {
 			break
 		}
-		typ = randgen.RandType(rng)
-		if typ.Family() == types.VoidFamily {
-			// Skip the void family because it doesn't have some basic
-			// comparison operators defined.
-			retry--
-		}
+		typ = randTyp()
 	}
 	return typ
 }

